oximeter/collector/src/agent.rs

```diff
@@ -166,35 +295,112 @@ async fn collection_task(
+    // are in-progress when we get an update to the producer's information. In
+    // that case, the collection is likely doomed, since the producer has moved
+    // and won't be available at the address the collection started with. This
+    // lets us restart that collection with the new information.
+    let (producer_info_tx, producer_info_rx) = watch::channel(producer);
+    let (forced_collection_tx, forced_collection_rx) =
+        mpsc::channel(N_QUEUED_FORCED_COLLECTIONS);
+    let (timer_collection_tx, timer_collection_rx) =
+        mpsc::channel(N_QUEUED_TIMER_COLLECTIONS);
+    let (result_tx, mut result_rx) = mpsc::channel(1);
+    tokio::task::spawn(inner_collection_loop(
+        log.clone(),
+        producer_info_rx,
+        forced_collection_rx,
+        timer_collection_rx,
+        result_tx,
+    ));
+
     loop {
         tokio::select! {
             message = inbox.recv() => {
                 match message {
                     None => {
-                        debug!(log, "collection task inbox closed, shutting down");
+                        debug!(
+                            log,
+                            "collection task inbox closed, shutting down"
+                        );
                         return;
                     }
                     Some(CollectionMessage::Shutdown) => {
-                        debug!(log, "collection task received shutdown request");
+                        debug!(
+                            log,
+                            "collection task received shutdown request"
+                        );
                         return;
                     },
                     Some(CollectionMessage::Collect(token)) => {
-                        debug!(log, "collection task received explicit request to collect");
-                        perform_collection(&log, &mut stats, &client, &producer, &outbox, Some(token)).await;
+                        debug!(
+                            log,
+                            "collection task received explicit request to collect"
+                        );
+                        match forced_collection_tx.try_send(token) {
+                            Ok(_) => trace!(
+                                log, "forwarded explicit request to collection task"
+                            ),
+                            Err(e) => {
+                                match e {
+                                    TrySendError::Closed(tok) => {
+                                        debug!(
+                                            log,
+                                            "collection task forced collection \
+                                            queue is closed. Attempting to \
+                                            notify caller and exiting.",
+                                        );
+                                        let _ = tok.send(Ok(()));
```

Comment:
Should this be some kind of error instead of `Ok(())`? In this case we definitely did not force a collection to happen, right?

Reply:
That's true, yeah. I've sort of been operating on the assumption that queues only close if we're shutting down anyway. But we can still send an error of some kind, I'll update this.

Reply:
Added an error variant for this case specifically in 69dfa7a9d.